src/julia_locks.h

```diff
@@ -22,10 +22,14 @@ JL_DLLEXPORT int _jl_mutex_trylock_nogc(jl_task_t *self, jl_mutex_t *lock) JL_NO
 JL_DLLEXPORT int _jl_mutex_trylock(jl_task_t *self, jl_mutex_t *lock);
 JL_DLLEXPORT void _jl_mutex_unlock(jl_task_t *self, jl_mutex_t *lock);
 JL_DLLEXPORT void _jl_mutex_unlock_nogc(jl_mutex_t *lock) JL_NOTSAFEPOINT;
+// Unfortunately we can't include `julia_internal.h` here, so we need to forward declare this
+JL_DLLEXPORT uint64_t jl_record_time_for_tls_metric(void) JL_NOTSAFEPOINT;
 
 static inline void jl_mutex_wait(jl_mutex_t *lock, int safepoint)
 {
+    uint64_t t0 = jl_record_time_for_tls_metric();
     _jl_mutex_wait(jl_current_task, lock, safepoint);
+    jl_increment_timing_tls_metric(jl_current_task->ptls, internal_spinlock_time, jl_record_time_for_tls_metric() - t0);
```

Comment:
Does this just capture the time waiting to acquire the lock? if so, don't we also need to time spent _holding_ certain important spin locks, namely the codegen lock?

Or, in other words: don't we need to account for `compilation_time` (and minus this out of user time)?

Reply:
> Does this just capture the time waiting to acquire the lock?

Yes.

> don't we need to account for compilation_time (and minus this out of user time)?

Possibly. But seems like the instrumentation to do this would be independent of the metric to measure time spinning on a lock.

I.e. seems like the codegen_lock is one of the few special locks for which is worth recording how long we held it, so it looks like we should instrument the places in which we acquire and release the codegen_lock itself, rather than doing it for all locks (since, as I mentioned before, we won't need information about how lock we held a lock in most cases).

Reply:
> we should instrument the places in which we acquire and release the codegen_lock itself

yeah, agreed

Reply:
Seems like we're already adding compilation time to `jl_cumulative_compile_time`, so this instrumentation is already present...